parity-crypto/src/secp256k1_alt.rs

```diff
@@ -0,0 +1,411 @@
+
+	fn to_vec(&self) -> Self::VecRepr {
+		self.0.serialize()[1..PUB_SIZE + 1].to_vec()
+	}
+
+	/// Should move to another trait.
+	fn to_compressed_vec(&self) -> Self::CompVecRepr {
+		self.0.serialize_compressed().to_vec()
+	}
+
+	fn verify(&self, signature: &[u8], message: &[u8]) -> Result<bool, Error> {
+		let mut buf = [0;32];
+		buf.copy_from_slice(&message[..]); // panic on incorrect message size
+		let message = Message::parse(&buf);
+		let mut buf = [0;64];
+		buf.copy_from_slice(&signature[..64]); // panic on incorrect signature size
+		let signature = Signature::parse(&buf);
+
+		Ok(secp256k1::verify(&message, &signature, &self.0))
+	}
+
+}
+
+impl SecretKeyTrait for SecretKey {
+	type VecRepr = ClearOnDrop<Vec<u8>>;
+
+	fn to_vec(&self) -> Self::VecRepr {
+		ClearOnDrop::new(self.0.serialize().to_vec())
+	}
+
+	fn sign(&self, message: &[u8]) -> Result<Vec<u8>, Error> {
+
+	 	let mut buf = [0;32];
+		buf.copy_from_slice(&message[..]); // panic on incorrect message size
+		let message = Message::parse(&buf);
+		let (sig, rec_id) = secp256k1::sign(&message, &self.0)?;
+		let mut data_arr = vec![0; 65];
+		data_arr[0..64].copy_from_slice(&sig.serialize());
+		data_arr[64] = rec_id.serialize();
+		Ok(data_arr)
+	}
+
+}
+
+/// random secret key for rand 0.5
+fn random_sec<R: Rng>(rng: &mut R) -> SecretKey {
+	loop {
+		let mut ret = [0u8; 32];
+		rng.fill_bytes(&mut ret);
+
+		match Secp256k1::secret_from_slice(&ret) {
+			Ok(key) => return key,
+			Err(_) => (),
+		}
+	}
+}
+
+impl FixAsymSharedSecret for SecretKey {
+	type Other = PublicKey;
+	type Result = SharedSecretAsRef;
+
+	fn shared_secret(&self, publ: &Self::Other) -> Result<Self::Result, Error> {
+		let shared = secp256k1::SharedSecret::new(&publ.0, &self.0)?;
+		Ok(SharedSecretAsRef(shared))
+	}
+
+}
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct SharedSecretAsRef(pub secp256k1::SharedSecret);
+
+impl AsRef<[u8]> for SharedSecretAsRef {
+	fn as_ref(&self) -> &[u8] {
+		self.0.as_ref()
+	}
+}
+
+fn aff_to_public(aff_pub: &mut Affine) -> Result<PublicKeyInner, Error> {
+	let mut buff = [4;65];
+	let mut buff2 = [0;32];
+	aff_pub.x.normalize();
+	aff_pub.x.fill_b32(&mut buff2);
+	buff[1..33].copy_from_slice(&buff2[..]);
+	aff_pub.y.normalize();
+	aff_pub.y.fill_b32(&mut buff2);
+	buff[33..65].copy_from_slice(&buff2[..]);
+	Ok(PublicKeyInner::parse(&buff)?)
+}
+
+struct SecretScalar(pub Scalar);
+
+impl Drop for SecretScalar {
+	fn drop(&mut self) {
+		self.0.clear();
+	}
+}
+
+/* private inner method but this would avoid a scalar instantiation
+fn mul_in_place_scalar(a: &mut Scalar, b: &Scalar) {
+	let mut l = [0u32; 16];
+	a.mul_512(b, &mut l);
+	a.reduce_512(&l);
+}
+*/
+
+impl FiniteField for Secp256k1 {
+
+	fn generator_x() -> &'static[u8] { &GENERATOR_X[..] }
+	fn generator_y() -> &'static[u8] { &GENERATOR_Y[..] }
+	fn curve_order() -> &'static[u8] { &CURVE_ORDER[..] }
+
+	fn public_mul(pub_key: &mut Self::PublicKey, sec_key: &Self::SecretKey) -> Result<(), Error> {
```

Comment:
Want to ask -- what are the use cases for `public_mul` / `public_add` / `secret_add` / `secret_mul` / `secret_inv`? If they're useful, I should implement them in `libsecp256k1`.

Reply:
It is mainly here for compatibility (the point of the crate is not the clean api/trait), so it is operation on key use by :
- secret_store for key calculation/distribution/verification.
- ethkey for all key derivation operations (hd wallet).

It would be certainly better if the lib got higher level operation (moving those low level key math (there is probably some redundancy between each crates math primitives) to a single parity crypto impl.

But to keep thing simple I use those trait.

Reply:
Got this as `PublicKey::tweak_add_assign`.

Reply:
Added this as `PublicKey::tweak_mul_assign`.

Reply:
> It is mainly here for compatibility

Maybe this PR is a good chance to add a deprecation notice?

Reply:
Would probably make sense in a second step; seems a bit premature now (moving those math usage is a bit of a refactoring and I am not sure I can assert that it is doable or just even wanted at this point).